src/client/dfuse/inval.c

```diff
@@ -0,0 +1,438 @@
+ * but at least 2 seconds and at most 60.
+ * As this relates to releasing resources there is no additional benefit in finer grained time
+ * control than this.
+ */
+
+/* Grace period before invalidating directories or non-directories.  Needs to be long enough so that
+ * entries in the working set are invalidated but short enough to be meaningful.
+ * Directories that are used as the cwd for processes can cause problems with being invalidated too
+ * early so use a higher value here.
+ */
+#define INVAL_DIRECTORY_GRACE (60 * 30)
+#define INVAL_FILE_GRACE      2
+
+/* Represents one timeout value (time).  Maintains a ordered list of dentries that are using
+ * this timeout.
+ */
+struct dfuse_time_entry {
+	d_list_t inode_list;
+	double   time;
+	d_list_t dte_list;
+};
+
+/* Core data structure, maintains a list of struct dfuse_time_entry lists.*/
+struct dfuse_ival {
+	d_list_t             time_entry_list;
+	struct fuse_session *session;
+	bool                 session_dead;
+};
+
+/* The core data from struct dfuse_inode_entry.  No additional inode references are held on inodes
+ * because of there place on invalidate lists, rather inodes are removed from any list on close.
+ * Therefore once a decision is made to evict an inode then a copy of the data is needed as once
+ * the ival_lock is dropped the inode could be freed.  This is not a problem if this happens as the
+ * kernel will simply return ENOENT.
+ */
+struct inode_core {
+	char       name[NAME_MAX + 1];
+	fuse_ino_t parent;
+	bool       dir;
+};
+
+/* Number of dentries to invalidate per iteration. This value affects how long the lock is held,
+ * after the invalidations happen then another iteration will start immediately.  Invalidation of
+ * directories however trigger many forget calls so we want to make use of this where possible so
+ * keep this batch size small.
+ */
+#define EVICT_COUNT 8
+
+static pthread_mutex_t   ival_lock = PTHREAD_MUTEX_INITIALIZER;
+static bool              ival_stop;
+static pthread_t         ival_thread;
+static sem_t             ival_sem;
+static struct dfuse_ival ival_data;
+
+/* Eviction loop, run periodically in it's own thread
+ *
+ * Returns true if there is more work to do.  If false then *sleep_time is set in seconds.
+ */
+static bool
+ival_loop(int *sleep_time)
+{
+	struct dfuse_time_entry *dte;
+	struct inode_core        ic[EVICT_COUNT] = {};
+	int                      idx             = 0;
+	double                   sleep           = (60 * 1) - 1;
+
+	D_MUTEX_LOCK(&ival_lock);
+
+	/* Walk the list, oldest first */
+	d_list_for_each_entry(dte, &ival_data.time_entry_list, dte_list) {
+		struct dfuse_inode_entry *inode, *inodep;
+
+		DFUSE_TRA_DEBUG(dte, "Iterating for timeout %lf", dte->time);
+
+		d_list_for_each_entry_safe(inode, inodep, &dte->inode_list, ie_evict_entry) {
+			double timeout;
+
+			if (dfuse_dentry_get_valid(inode, dte->time, &timeout)) {
+				DFUSE_TRA_DEBUG(inode, "Keeping left %lf " DF_DE, timeout,
```

Comment:
I think this will log at most once per timeout value per thread wakeup which itself is a minimum of two seconds.  Debug logs getting out of control is something than can happen but I don't think this one will be particularly noisy.